// angular-pipes/esm2015/math/bytes.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Pipe } from '@angular/core';
import { isNumberFinite, isPositive, isInteger, toDecimal } from '../utils/utils';
export class BytesPipe {
    /**
     * @param {?} input
     * @param {?=} decimal
     * @param {?=} from
     * @param {?=} to
     * @return {?}
     */
    transform(input, decimal = 0, from = 'B', to) {
        if (!(isNumberFinite(input) &&
            isNumberFinite(decimal) &&
            isInteger(decimal) &&
            isPositive(decimal))) {
            return input;
        }
        let /** @type {?} */ bytes = input;
        let /** @type {?} */ unit = from;
        while (unit !== 'B') {
            bytes *= 1024;
            unit = /** @type {?} */ ((BytesPipe.formats[unit].prev));
        }
        if (to) {
            const /** @type {?} */ format = BytesPipe.formats[to];
            const /** @type {?} */ result = toDecimal(BytesPipe.calculateResult(format, bytes), decimal);
            return BytesPipe.formatResult(result, to);
        }
        for (const /** @type {?} */ key in BytesPipe.formats) {
            const /** @type {?} */ format = BytesPipe.formats[key];
            if (bytes < format.max) {
                const /** @type {?} */ result = toDecimal(BytesPipe.calculateResult(format, bytes), decimal);
                return BytesPipe.formatResult(result, key);
            }
        }
    }
    /**
     * @param {?} result
     * @param {?} unit
     * @return {?}
     */
    static formatResult(result, unit) {
        return `${result} ${unit}`;
    }
    /**
     * @param {?} format
     * @param {?} bytes
     * @return {?}
     */
    static calculateResult(format, bytes) {
        const /** @type {?} */ prev = format.prev ? BytesPipe.formats[format.prev] : undefined;
        return prev ? bytes / prev.max : bytes;
    }
}
BytesPipe.formats = {
    'B': { max: 1024 },
    'KB': { max: Math.pow(1024, 2), prev: 'B' },
    'MB': { max: Math.pow(1024, 3), prev: 'KB' },
    'GB': { max: Math.pow(1024, 4), prev: 'MB' },
    'TB': { max: Number.MAX_SAFE_INTEGER, prev: 'GB' }
};
BytesPipe.decorators = [
    { type: Pipe, args: [{
                name: 'bytes'
            },] },
];
function BytesPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    BytesPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    BytesPipe.ctorParameters;
    /** @type {?} */
    BytesPipe.formats;
}
